src/handlers/RebalanceHLPHandler.sol

```diff
@@ -0,0 +1,167 @@
+
+  uint16 public constant BPS = 100_00;
+
+  IRebalanceHLPService public service;
+  IVaultStorage public vaultStorage;
+  ICalculator public calculator;
+
+  IERC20Upgradeable public sglp;
+
+  uint16 public minExecutionFeeBPS;
+
+  mapping(address => bool) public whitelistExecutors;
+
+  event LogSetWhitelistExecutor(address indexed _account, bool _active);
+  event LogSetMinExecutionFee(uint16 _oldFee, uint16 _newFee);
+  event LogSetRebalanceHLPService(address indexed _oldService, address indexed _newService);
+
+  modifier onlyWhitelisted() {
+    // if not whitelist
+    if (!whitelistExecutors[msg.sender]) {
+      revert RebalanceHLPHandler_OnlyWhitelisted();
+    }
+    _;
+  }
+
+  function initialize(
+    address _rebalanceHLPService,
+    address _calculator,
+    uint16 _minExecutionFeeBPS
+  ) external initializer {
+    __Ownable_init();
+    __ReentrancyGuard_init();
+    // gas opt
+    IRebalanceHLPService _service = IRebalanceHLPService(_rebalanceHLPService);
+    service = _service;
+    vaultStorage = _service.vaultStorage();
+    sglp = _service.sglp();
+    calculator = ICalculator(_calculator);
+    minExecutionFeeBPS = _minExecutionFeeBPS;
+  }
+
+  function setWhiteListExecutor(address _executor, bool _active) external onlyOwner {
+    if (_executor == address(0)) {
+      revert RebalanceHLPHandler_AddressIsZero();
+    }
+    whitelistExecutors[_executor] = _active;
+    emit LogSetWhitelistExecutor(_executor, _active);
+  }
+
+  function setMinExecutionFeeBPS(uint16 _newExecutionFeeBPS) external onlyOwner {
+    if (_newExecutionFeeBPS == 0) {
+      revert RebalanceHLPHandler_AmountIsZero();
+    }
+    emit LogSetMinExecutionFee(minExecutionFeeBPS, _newExecutionFeeBPS);
+    minExecutionFeeBPS = _newExecutionFeeBPS;
+  }
+
+  function setRebalanceHLPService(address _newService) external nonReentrant onlyOwner {
+    if (_newService == address(0)) {
+      revert RebalanceHLPHandler_AddressIsZero();
+    }
+    emit LogSetRebalanceHLPService(address(service), _newService);
+    service = IRebalanceHLPService(_newService);
+  }
+
+  function executeLogicReinvestNonHLP(
+    IRebalanceHLPService.ExecuteReinvestParams[] calldata _params
+  ) external nonReentrant onlyWhitelisted returns (uint256 receivedGlp) {
+    if (_params.length == 0) revert RebalanceHLPHandler_ParamsIsEmpty();
+    _validateReinvestInput(_params);
+
+    // Get current HLP value
+    uint256 totalHlpValueBefore = calculator.getHLPValueE30(true);
+    // Execute logic at Service
+    receivedGlp = service.executReinvestNonHLP(_params);
+    _validateHLPValue(totalHlpValueBefore);
+  }
+
+  function executeLogicWithdrawGLP(
+    IRebalanceHLPService.ExecuteWithdrawParams[] calldata _params
+  ) external nonReentrant onlyWhitelisted returns (IRebalanceHLPService.WithdrawGLPResult[] memory result) {
+    if (_params.length == 0) revert RebalanceHLPHandler_ParamsIsEmpty();
+    _validateWithdrawInput(_params);
+    // Get current HLP value
```

Comment:
We will need to update the price in this function, because we are checking TVL and we want the most up to date price.

```
IEcoPyth(pyth).updatePriceFeeds(_priceData, _publishTimeData, _minPublishTime, _encodedVaas);
```
See `CrossMarginHandler.executeOrder()` for referrence.